cpp/util/etcd.cc

```diff
@@ -97,12 +83,73 @@ util::error::Code ErrorCodeForHttpResponseCode(int response_code) {
 }
 
 
-Status StatusFromResponseCode(const int response_code,
-                              const shared_ptr<JsonObject>& json) {
+util::error::Code StatusCodeFromEtcdErrorCode(int etcd_code) {
+  switch (etcd_code) {
+    case 100:  // Key not found
+      return util::error::NOT_FOUND;
+    case 101:  // Compare failed
+    case 102:  // Not a file
+    case 104:  // Not a directory
+    case 105:  // Key already exists
+    case 107:  // Root is read-only
+    case 108:  // Directory not empty
+      return util::error::FAILED_PRECONDITION;
+
+    case 201:  // PrevValue missing
+    case 202:  // Provided TTL is not a number
+    case 203:  // Provided index is not a number
+    case 209:  // Invalid field
+    case 210:  // Invalid POST form
+      return util::error::FAILED_PRECONDITION;
+
+    case 300:  // Raft Internal Error
+    case 301:  // During Leader Election
+      return util::error::UNAVAILABLE;
+
+    case 400:  // Watcher is cleared due to etcd recovery
+    case 401:  // Event in requested index is outdated and cleared.
+      return util::error::INVALID_ARGUMENT;
+
+    default:
+      return util::error::UNKNOWN;
+  }
+}
+
+
+string EtcdErrorMessage(const shared_ptr<JsonObject>& json) {
+  const JsonString message(*json, "message");
+  const JsonString cause(*json, "cause");
+
+  string ret("Etcd message: ");
+  if (message.Ok()) {
+    ret += message.Value();
+    if (cause.Ok()) {
+      ret += ", cause: " + string(cause.Value());
+    }
+  } else {
+    ret = json->DebugString();
+  }
+  return ret;
+}
+
+
+Status StatusFromResponse(const int response_code,
+                          const shared_ptr<JsonObject>& json) {
+  // Prefer the etcd errorCode if there is one:
+  if (json->Ok()) {
```

Comment:
This would bomb if `json` is null, maybe `StatusFromResponse` should take a `const JsonObject&`?